addons/packages/contour/1.20.1/test/unittest/contour_test.go

```diff
@@ -0,0 +1,729 @@
+			Expect(err).To(ContainSubstring("certificates.renewBefore must be provided when certificates.useCertManager is true"))
+		})
+	})
+
+	Context("Envoy host ports enabled, no http port", func() {
+		BeforeEach(func() {
+			values = noHostPortHttp
+		})
+
+		It("renders with an error", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(ContainSubstring("envoy.hostPorts.http must be provided when envoy.hostPorts.enable is true"))
+		})
+	})
+
+	Context("Envoy host ports enabled, no https port", func() {
+		BeforeEach(func() {
+			values = noHostPortHttps
+		})
+
+		It("renders with an error", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(ContainSubstring("envoy.hostPorts.https must be provided when envoy.hostPorts.enable is true"))
+		})
+	})
+
+	Context("Invalid envoy log level", func() {
+		BeforeEach(func() {
+			values = invalidEnvoyLogLevel
+		})
+
+		It("renders with an error", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(ContainSubstring("envoy.logLevel must be one of trace|debug|info|warning/warn|error|critical|off"))
+		})
+	})
+
+	Context("No envoy termination grace period seconds", func() {
+		BeforeEach(func() {
+			values = noTerminationGracePeriodSeconds
+		})
+
+		It("renders with an error", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(ContainSubstring("envoy.terminationGracePeriodSeconds must be provided"))
+		})
+	})
+
+	Context("Invalid envoy service type", func() {
+		BeforeEach(func() {
+			values = invalidEnvoyServiceType
+		})
+
+		It("renders with an error", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(ContainSubstring("envoy.service.type must be either LoadBalancer or NodePort or ClusterIP"))
+		})
+	})
+
+	Context("Invalid envoy service external traffic policy", func() {
+		BeforeEach(func() {
+			values = invalidEnvoyServiceExternalTrafficPolicy
+		})
+
+		It("renders with an error", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(ContainSubstring("envoy.service.externalTrafficPolicy must be either Cluster or Local"))
+		})
+	})
+
+	Context("Invalid AWS load balancer type", func() {
+		BeforeEach(func() {
+			values = invalidAwsLoadBalancerType
+		})
+
+		It("renders with an error", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(ContainSubstring("envoy.service.aws.loadBalancerType must be either classic or nlb when infrastructureProvider is aws"))
+		})
+	})
+	// END validation tests
+
+	// START Envoy service defaulting
+	Context("Envoy service defaults when no infrastructure provider is specified", func() {
+		BeforeEach(func() {
+			values = ""
+		})
+
+		It("defaults the Envoy service type to LoadBalancer", func() {
+			Expect(err).NotTo(HaveOccurred())
+
+			docs := findDocsContainingLines(output, "kind: Service", "name: envoy")
+			Expect(docs).To(HaveLen(1), "Envoy service not found")
+
+			service := unmarshalService(docs[0])
+			Expect(service.Spec.Type).To(Equal(corev1.ServiceTypeLoadBalancer))
```

Comment:
I opted against checking the entire ~10k lines of YAML for each test case as I don't think that's manageable on an ongoing basis since they'd all have to be updated regularly as things upstream change. I am thinking about adding a small number of test cases that do check the full YAML, though, as a guard against any weird regressions in the overlays. Can be done in a separate PR since this one's big enough already, but interested in others' thoughts on the approach here.

Reply:
At first, I didn't like the testing approach because it felt very string dependent instead of YAML native.
After reading through it more and understanding the test-helpers, I think it's pretty precise and minimizes the need to Unmarshal a bunch of YAML which should be nice and fast.

I'm not understanding what you're saying about checking the 10k lines of YAML for each test case.
As long as ytt is returning successfully, that's a pretty good check for the YAML build as a whole with the input values.

I agree with the TODO's to assert the outputs for each infra provider.

Running the tests is a good experience!

Reviewing them all in a single, long file in GitHub is unpleasant, but I'm guessing future diffs will be easier to look through.
I can't think of a way to split up the different sections of tests into separate files without redefining more `Describe` blocks or using more functions and creating unnecessary indirection.

It might be helpful to move the Makefile up a directory and also add the `ytt`/`kbld`/`imgpkg push` commands related to pinning image versions, checking ytt builds, publishing, and other package maintenance needs.

Reply:
> I'm not understanding what you're saying about checking the 10k lines of YAML for each test case.
As long as ytt is returning successfully, that's a pretty good check for the YAML build as a whole with the input values.

I have the tests verifying specific properties of the YAML (e.g. the Envoy service type, the Contour deployment args, etc) rather than having an entire 10k line of "expected YAML" for each test case. That leaves the door open for overlay logic to be inadvertently altering some other portion of the YAML output, without these tests detecting it. If you look at e.g. [external-dns tests](https://github.com/vmware-tanzu/community-edition/tree/main/addons/packages/external-dns/0.10.0/test/unittest/fixtures/expected), they use files containing the entire expected YAML output for each test case. However, their YAML is orders of magnitude shorter than ours.

> Reviewing them all in a single, long file in GitHub is unpleasant, but I'm guessing future diffs will be easier to look through.

Yeah, future reviews should be easier since there should be minimal changes/additions going forward.

> It might be helpful to move the Makefile up a directory and also add the ytt/kbld/imgpkg push commands related to pinning image versions, checking ytt builds, publishing, and other package maintenance needs.
It would make the learning curve easier for users who have a sudden need to fork or contribute fixes.

I'm pretty sure this Makefile needs to be where it is to integrate with repo-wide tooling for running package tests. We could look at adding another one for those other tasks you're talking about, although much of that is already covered in the repo-level Makefile (e.g. `make vendir-sync-package`, `make lock-package-images`, `make push-package`).

Reply:
yeah maybe we can do something around pulling the CRD part of the expected YAML out into some common helper etc. (or something to that effect) so we don't have to keep repeating them if/when we do add holistic tests, they should be static regardless of config changes and are the longest section line-wise of our base YAML

Reply:
Yeah that's a good idea, they probably are also the YAML that tends to change the most upstream on a release-to-release basis (not that other YAMLs can't change, just that they're less frequent so maybe less of a maintenance burden if we do want to fully validate them).